File: russian_docs_ocr/document_processing/pipeline_modules/doctype_angles_classificator/doctype_angles.py
from ..base_module import BaseModule
from typing import Union
from pathlib import Path
import numpy as np
import cv2

class DocTypeAngles(BaseModule):
    """Classifies document type and 90-degree rotation angle from an image.

    Identifies whether the image contains a passport, driver's license etc.
    and by how many degrees (multiple of 90) it is rotated.

    Output follows the standard module contract - a single dict keyed by
    ``model_name`` ('DocTypeAngles') with a flat payload:

        {'DocTypeAngles': {
            'doc_type': str,                # e.g. 'INTPASSPORT_2011', 'NONE'
            'doc_type_confidence': float,   # 1 - dist/threshold, rounded
            'angle': int,                   # 0/90/180/270
            'angle_confidence': float,
            'warped_img': np.ndarray,       # predict_transform only: upright image
        }}
    """

    def __init__(self, model_format: str = 'ONNX', device='cpu', verbose: bool = False):
        """Initializes document type and angles detection model."""
        self.model_name = 'DocTypeAngles'
        super().__init__(self.model_name, model_format=model_format, device=device, verbose=verbose)

    def __predict_meta(self, img: np.ndarray) -> dict:
        """Runs the model and builds the flat payload dict (no rotation)."""
        doctype_meta, angle_meta = self.model.predict(img)
        doc_type, dist, thresh = doctype_meta
        angle, angle_conf = angle_meta
        # thresh == 0 is the postprocessing's "no centroid within radius" sentinel
        # (it also sets dist to inf, so the division would raise ZeroDivisionError).
        # That case is maximally-unknown, hence confidence 0.0.
        confidence = np.round(1 - dist / thresh, 2) if thresh > 0 else 0.0
        return {
            'doc_type': doc_type,
            'doc_type_confidence': confidence,
            'angle': angle,
            'angle_confidence': angle_conf,
        }

    def predict(self, img: Union[str, Path, np.ndarray]) -> dict:
        """Predicts document type and rotation angle with confidences.

        Args:
            img: Input document image

        Returns:
            {'DocTypeAngles': {...}} - see class docstring.
        """
        img = self.load_img(img)
        return {self.model_name: self.__predict_meta(img)}

    def predict_transform(self, img: Union[str, Path, np.ndarray]) -> dict:
        """Same as predict(), plus the image rotated upright ('warped_img').

        Args:
            img: Input document image

        Returns:
            {'DocTypeAngles': {...}} - see class docstring.
        """
        img = self.load_img(img)
        meta = self.__predict_meta(img)
        for _ in range(meta['angle'] // 90):
            img = cv2.rotate(img, cv2.ROTATE_90_COUNTERCLOCKWISE)
        meta['warped_img'] = img
        return {self.model_name: meta}

File: ports/go/internal/docproc/modules/doctypeangles.go
// Package modules holds one thin type per ML module, mirroring pipeline_modules/.
package modules

import (
	"fmt"

	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/config"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/imaging"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/inference"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/models"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/postprocess"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/tensor"
)

// DocTypeAngles classifies the document type and its 90-degree rotation.
// Port of pipeline_modules/doctype_angles_classificator/doctype_angles.py.
type DocTypeAngles struct {
	model *models.Model
	// angleLabels are the numeric labels of the second head, [0,90,180,270]. Kept
	// separately because MultiClass hands back the label as a STRING and the caller
	// needs the integer to decide how many times to rotate.
	angleLabels []int
}

// DocTypeResult is the module's flat payload, matching the dict the Python module
// returns under the key 'DocTypeAngles'.
//
// Field names are the wire names on purpose: this payload is emitted as the
// `doctype.label` conformance stage and compared against the Python golden, so the
// JSON tags are part of the contract (conformance/spec/stages.md).
type DocTypeResult struct {
	DocType           string  `json:"doc_type"`
	DocTypeConfidence float64 `json:"doc_type_confidence"`
	Angle             int     `json:"angle"`
	AngleConfidence   float64 `json:"angle_confidence"`
}

const moduleName = "DocTypeAngles"

// NewDocTypeAngles loads the module from the repository's model directory.
func NewDocTypeAngles(root string, paths *config.ModelPaths, format string,
	device inference.Device, threads int) (*DocTypeAngles, error) {

	dir, err := paths.Dir(moduleName, format)
	if err != nil {
		return nil, err
	}
	m, err := models.Load(root, dir, device, threads)
	if err != nil {
		return nil, fmt.Errorf("modules: %s: %w", moduleName, err)
	}
	if len(m.Config.Outputs) != 2 {
		_ = m.Close()
		return nil, fmt.Errorf("modules: %s expects 2 outputs (embeddings, angle), got %d",
			moduleName, len(m.Config.Outputs))
	}
	angleLabels, err := m.Config.Outputs[1].LabelsAsInts()
	if err != nil {
		_ = m.Close()
		return nil, fmt.Errorf("modules: %s angle labels: %w", moduleName, err)
	}
	return &DocTypeAngles{model: m, angleLabels: angleLabels}, nil
}

func (d *DocTypeAngles) Close() error { return d.model.Close() }

// Predict returns the type and angle without rotating anything.
func (d *DocTypeAngles) Predict(img imaging.Image) (DocTypeResult, error) {
	out, err := d.model.Predict(img)
	if err != nil {
		return DocTypeResult{}, err
	}

	// ONE checked assertion per output, at the layer that knows what it asked for.
	// This is the alternative to threading `any` through every caller — see
	// CONVENTIONS §5.
	metric, ok := out[0].(postprocess.MetricResult)
	if !ok {
		return DocTypeResult{}, fmt.Errorf("modules: %s output 0 is %T, want MetricResult",
			moduleName, out[0])
	}
	angle, ok := out[1].(postprocess.ClassResult)
	if !ok {
		return DocTypeResult{}, fmt.Errorf("modules: %s output 1 is %T, want ClassResult",
			moduleName, out[1])
	}

	// confidence = round(1 - dist/threshold, 2), and 0.0 when threshold is the
	// "no centroid within radius" sentinel — dividing by it would be a
	// ZeroDivisionError in Python and +Inf here, and the case is maximally unknown
	// anyway (doctype_angles.py:38).
	//
	// RoundHalfEven, not math.Round: np.round breaks ties to EVEN, and two decimals
	// on a value derived from a float distance reaches ties in practice.
	confidence := 0.0
	if metric.Threshold > 0 {
		confidence = tensor.RoundHalfEven(1-metric.Distance/metric.Threshold, 2)
	}

	angleDeg, err := d.angleFromLabel(angle.Label)
	if err != nil {
		return DocTypeResult{}, err
	}

	return DocTypeResult{
		DocType:           metric.Label,
		DocTypeConfidence: confidence,
		Angle:             angleDeg,
		AngleConfidence:   float64(angle.Confidence),
	}, nil
}

// PredictTransform is Predict plus the image rotated upright.
//
// `angle // 90` counter-clockwise rotations, exactly as the reference does. The angle
// is a multiple of 90 by construction (the head's labels are 0/90/180/270), so this is
// 0-3 whole-image rotations and never an interpolating warp.
func (d *DocTypeAngles) PredictTransform(img imaging.Image) (DocTypeResult, imaging.Image, error) {
	meta, err := d.Predict(img)
	if err != nil {
		return DocTypeResult{}, imaging.Image{}, err
	}

	current := img.Clone()
	for i := 0; i < meta.Angle/90; i++ {
		next := imaging.Rotate90CCW(current)
		_ = current.Close()
		current = next
	}
	return meta, current, nil
}

// angleFromLabel maps the winning class back to degrees.
//
// MultiClass returns the label as a string because most heads have textual classes;
// this head's are numeric, so the string is matched against the declared labels rather
// than parsed. Matching beats parsing here: it fails loudly if the config and the
// model ever disagree, instead of silently producing a plausible number.
func (d *DocTypeAngles) angleFromLabel(label string) (int, error) {
	for _, v := range d.angleLabels {
		if fmt.Sprintf("%g", float64(v)) == label || fmt.Sprintf("%d", v) == label {
			return v, nil
		}
	}
	return 0, fmt.Errorf("modules: %s angle label %q is not one of %v",
		moduleName, label, d.angleLabels)
}
